From 8f9cec845b9449ab8e9256093ebec9ae420b67b6 Mon Sep 17 00:00:00 2001
From: Aaradhana Sahu <quic_aarasahu@quicinc.com>
Date: Mon, 23 Dec 2024 14:17:55 +0530
Subject: [PATCH] UPSTREAM:PROTOCOL: wifi: mac80211: Fill maximum number of simultaneous links subfield in association request frame

Currently, the maximum number of simultaneous links subfield in the
basic multi-link element is not filled in association request frame,
resulting in maximum number of simultaneous links subfield being zero.

As per IEEE std 802.11be-2024 section 35.3.16.2, if a multi-link
radio non-AP MLD requests more than one link during multi-link setup,
the non-AP MLD shall set the maximum number of simultaneous links
subfield in the basic multi-link element in transmitted Management
frames to a value of 1 or greater.

Therefore, fill simultaneous links subfield in the association request
management frame with value N-1, where N is the number of active link.

Patch-dependency: None
Patch-work:
[PATCH] wifi: mac80211: Fill maximum number of simultaneous links subfield for MLD station

Signed-off-by: Aaradhana Sahu <quic_aarasahu@quicinc.com>
---
 net/mac80211/mlme.c | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/net/mac80211/mlme.c b/net/mac80211/mlme.c
index 710c659..202d16b 100644
--- a/net/mac80211/mlme.c
+++ b/net/mac80211/mlme.c
@@ -1357,7 +1357,7 @@ static void ieee80211_assoc_add_ml_elem(struct ieee80211_sub_if_data *sdata,
 	const struct wiphy_iftype_ext_capab *ift_ext_capa;
 	__le16 eml_capa = 0, mld_capa_ops = 0;
 	unsigned int link_id;
-	u8 *ml_elem_len;
+	u8 *ml_elem_len, max_simul_links = 0;
 	void *capab_pos;
 
 	if (!ieee80211_vif_is_mld(&sdata->vif))
@@ -1391,6 +1391,17 @@ static void ieee80211_assoc_add_ml_elem(struct ieee80211_sub_if_data *sdata,
 			cpu_to_le16(IEEE80211_MLC_BASIC_PRES_EML_CAPA);
 		skb_put_data(skb, &eml_capa, sizeof(eml_capa));
 	}
+
+	for (link_id = 0; link_id < IEEE80211_MLD_MAX_NUM_LINKS; link_id++) {
+		if (!assoc_data->link[link_id].bss ||
+		    link_id == assoc_data->assoc_link_id)
+			continue;
+
+		max_simul_links++;
+	}
+
+	mld_capa_ops |= cpu_to_le16(max_simul_links & IEEE80211_MLD_CAP_OP_MAX_SIMUL_LINKS);
+
 	/* need indication from userspace to support this */
 	mld_capa_ops &= ~cpu_to_le16(IEEE80211_MLD_CAP_OP_TID_TO_LINK_MAP_NEG_SUPP);
 	skb_put_data(skb, &mld_capa_ops, sizeof(mld_capa_ops));
-- 
2.34.1

